### ISSUE TITLE
[optimization]reproduce the result of paper "ADAPTIVE FEDERATED OPTIMIZATION"

### ISSUE THREAD

**Issue by Jiaxiangren** [google-research/federated#48]
Hi, there,

I am trying to reproduce the FedAdam of cifar100 in figure 1 by using the command below:

bazel run :trainer -- --task=cifar100_image --total_rounds=4000 --client_optimizer=sgd --client_learning_rate=0.031622 --client_batch_size=20 --server_optimizer=adam --server_learning_rate=1.0 --clients_per_round=10 --client_epochs_per_round=1 --experiment_name=fed_Adaptive_sgd_adam_opt -server_adam_epsilon=0.1 --client_datasets_random_seed=1

I got val accuracy around 45% but expected 53% in the paper.

Did I miss something there?
Thanks in advance for your help!

**Comment by Jiaxiangren** [comment]
[hparams.csv](https://github.com/google-research/federated/files/7524352/hparams.csv
[experiment.metrics.csv](https://github.com/google-research/federated/files/7524354/experiment.metrics.csv)
)

**Comment by Jiaxiangren** [comment]
The files below are my training log and haparameter settings that generated by program.
I found out at the end of training, I got large training acc around 85%, but low val acc around 45%.

**Comment by zcharles8** [comment]
Hi @Jiaxiangren. There are a couple of things to note:

1. If you look at Figure 6 of https://arxiv.org/pdf/2003.00295.pdf, you'll see that the best learning rate combination is at `--`--server_learning_rate=0.01 --client_learning_rate=0.1`.
2. Unfortunately, I believe that `tf.keras` has changed the behavior of random model initialization since I wrote this code. This can certainly affect accuracy, so in subsequent works I've found that averaging over 5 random trials is much more effective.
3. This averaging across 5 random trials is better yet when we also vary `--client_datasets_random_seed` over those 5 trials.

Essentially, getting exact reproducibility is unfortunately difficult due to extra sources of randomness right now. That being said, I would encourage you to consider "Adaptive Federated Optimization" as a comparative study, not an absolute one. The thing we consider important is the differences between optimizers.

Finally, if you want to reduce the generalization gap in CIFAR-100, you'll probably need to use some of the more common data augmentation strategies, as well as things like weight decay. These were not used in our experiments for simplicity.

**Comment by Jiaxiangren** [comment]
Hi, @zcharles8 , Thanks for your time.
I still got two questions,
1.For FedAdam of Cifar-100, based on the table 8 in the paper, the learning rate for client and server are 10^-1.5 and 10^0 respectively, is there any updated version of this paper?
2. you talk about running multiple trials is better, does that mean running multiple trials with same dataset seed or running multiple trials that each trial has different dataset seed?

thanks in advance for replying!

**Comment by zcharles8** [comment]
I think this is all dependent on the `epsilon` parameter. It could be that the values reported in Table 8 were not the best ones possible (I would have to re-check the experiments to make sure). I might recommend setting `epsilon = 1e-3` and using the results in Figure 6.

That all being said, I think the issues with TF changing random initialization and randomness may be more pressing. More recent work has made things a lot more robust by using `tff.learning.robust_aggregator` (derived from from https://arxiv.org/abs/1905.03871). If you're interested, https://arxiv.org/pdf/2106.07820.pdf, Appendix A talks in much more detail about hyperparameters, and exactly what was used for all plots in the paper. This also used 5 random trials (using different sampling seeds) to make things more robust.

Hopefully that helps.

**Comment by zcharles8** [comment]
It's worth noting that https://arxiv.org/pdf/2106.07820.pdf also suggests that `--server_learning_rate=0.001 --client_learning_rate=0.1` is effective for FedAdam on CIFAR-100. That being said, the paper uses `tff.learning.robust_aggregator`, so it might be a good idea to incorporate this (this can be done by calling something like
```
iterative_process = tff.learning.build_federated_averaging_process(
  ...,
  model_update_aggregation_factory=tff.learning.robust_aggregator()
)

**Comment by Jiaxiangren** [comment]
Hi, @zcharles8 , thanks for your suggestions.

I will close this.
